providers/ms_entra_id.go

```diff
@@ -0,0 +1,188 @@
+package providers
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"regexp"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/logger"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/requests"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/util"
+	"github.com/spf13/cast"
+)
+
+// MicrosoftEntraIDProvider represents provider for Azure Entra Authentication V2 endpoint
+type MicrosoftEntraIDProvider struct {
+	*OIDCProvider
+	multiTenantAllowedTenants []string
+
+	microsoftGraphURL *url.URL
+}
+
+const (
+	microsoftEntraIDProviderName = "Microsoft Entra ID"
+)
+
+var (
+	microsoftGraphURL = &url.URL{
+		Scheme: "https",
+		Host:   "graph.microsoft.com",
+		Path:   "/v1.0/me",
+	}
+)
+
+// NewMicrosoftEntraIDProvider initiates a new MicrosoftEntraIDProvider
+func NewMicrosoftEntraIDProvider(p *ProviderData, opts options.Provider) *MicrosoftEntraIDProvider {
+	p.setProviderDefaults(providerDefaults{
+		name: microsoftEntraIDProviderName,
+	})
+
+	return &MicrosoftEntraIDProvider{
+		OIDCProvider: NewOIDCProvider(p, opts.OIDCConfig),
+
+		multiTenantAllowedTenants: opts.MicrosoftEntraIDConfig.AllowedTenants,
+		microsoftGraphURL:         microsoftGraphURL,
+	}
+}
+
+// EnrichSession checks for group overage after calling generic EnrichSession
+func (p *MicrosoftEntraIDProvider) EnrichSession(ctx context.Context, session *sessions.SessionState) error {
+	if err := p.OIDCProvider.EnrichSession(ctx, session); err != nil {
+		return fmt.Errorf("unable to enrich session: %v", err)
+	}
+
+	hasGroupOverage, err := p.checkGroupOverage(session)
+	if err != nil {
+		return fmt.Errorf("unable to check token: %v", err)
+	}
+
+	if hasGroupOverage {
+		logger.Printf("entra overage found, reading groups from Graph API")
```

Comment:
Do we really want to log this?

Reply:
As this is an exceptional case, I vote for keeping it.